common/rpc/interceptor/namespace_validator.go

```diff
@@ -0,0 +1,142 @@
+)
+
+type (
+	// NamespaceValidatorInterceptor validates:
+	// 1. Namespace is specified in request if there is a `namespace` field.
+	// 2. Namespace exists.
+	// 3. Namespace is in correct state.
+	NamespaceValidatorInterceptor struct {
+		namespaceRegistry namespace.Registry
+	}
+)
+
+var (
+	ErrNamespaceNotSet              = serviceerror.NewInvalidArgument("Namespace not set on request.")
+	errInvalidNamespaceStateMessage = "Namespace has invalid state: %s. Must be %s."
+
+	// By default, enumspb.NAMESPACE_STATE_REGISTERED and enumspb.NAMESPACE_STATE_DEPRECATED are allowed states for all APIs that have `namespace` field in the request object.
+	allowedNamespaceStates = map[string][]enumspb.NamespaceState{
+		"StartWorkflowExecution":           {enumspb.NAMESPACE_STATE_REGISTERED},
+		"SignalWithStartWorkflowExecution": {enumspb.NAMESPACE_STATE_REGISTERED},
+		"DescribeNamespace":                {enumspb.NAMESPACE_STATE_REGISTERED, enumspb.NAMESPACE_STATE_DEPRECATED, enumspb.NAMESPACE_STATE_DELETED},
+	}
+)
+
+var _ grpc.UnaryServerInterceptor = (*NamespaceValidatorInterceptor)(nil).Intercept
+
+func NewNamespaceValidatorInterceptor(
+	namespaceRegistry namespace.Registry,
+) *NamespaceValidatorInterceptor {
+	return &NamespaceValidatorInterceptor{
+		namespaceRegistry: namespaceRegistry,
+	}
+}
+
+func (ni *NamespaceValidatorInterceptor) Intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	namespaceEntry, err := ni.extractNamespace(req)
+	if err != nil {
+		return nil, err
+	}
+	if namespaceEntry != nil {
+		_, methodName := splitMethodName(info.FullMethod)
+		allowedStates, allowedStatesDefined := allowedNamespaceStates[methodName]
+		if !allowedStatesDefined {
+			// If not explicitly defined, only enumspb.NAMESPACE_STATE_REGISTERED and enumspb.NAMESPACE_STATE_DEPRECATED are allowed.
+			if namespaceEntry.State() != enumspb.NAMESPACE_STATE_REGISTERED && namespaceEntry.State() != enumspb.NAMESPACE_STATE_DEPRECATED {
+				return nil, serviceerror.NewInvalidArgument(fmt.Sprintf(errInvalidNamespaceStateMessage, namespaceEntry.State(), fmt.Sprintf("%s or %s", enumspb.NAMESPACE_STATE_REGISTERED.String(), enumspb.NAMESPACE_STATE_DEPRECATED.String())))
+			}
+		} else {
+			isStateAllowed := false
+			for _, allowedState := range allowedStates {
+				if allowedState == namespaceEntry.State() {
+					isStateAllowed = true
+					break
+				}
+			}
+			if !isStateAllowed {
+				var allowedStatesStr []string
+				for _, allowedState := range allowedStates {
+					allowedStatesStr = append(allowedStatesStr, allowedState.String())
+				}
+				return nil, serviceerror.NewInvalidArgument(fmt.Sprintf(errInvalidNamespaceStateMessage, namespaceEntry.State(), strings.Join(allowedStatesStr, " or ")))
+			}
+		}
+	}
+
+	return handler(ctx, req)
+}
+
+func (ni *NamespaceValidatorInterceptor) extractNamespace(req interface{}) (*namespace.Namespace, error) {
```

Comment:
could you make use of https://github.com/temporalio/temporal/blob/13082f322d4a9e1a03188bf4f244abb78f867f1d/common/rpc/interceptor/namespace.go#L43

Reply:
Unfortunately, no. I need to distinguish cases when there is no `namespace` field on request and then it is not set.